test/jdk/java/security/Security/ProviderFiltering.java

```diff
@@ -0,0 +1,146 @@
+        System.out.println("Filter: " + filter);
+        System.out.println("Expected Provider(s): " +
+                (expectedPNs.length > 0 ? Arrays.toString(expectedPNs) :
+                "<NONE>"));
+        Provider ps[];
+        if (filter instanceof String filterStr) {
+            ps = Security.getProviders(filterStr);
+        } else if (filter instanceof Map filterMap) {
+            ps = Security.getProviders(filterMap);
+        } else {
+            throw new RuntimeException("Error: unknown input type: " + filter);
+        }
+
+        if (ps == null) {
+            if (expectedPNs.length != 0) {
+                throw new RuntimeException("Fail: expected provider(s) " +
+                        "not found");
+            }
+        } else {
+            if (ps.length == expectedPNs.length) {
+                // check the provider names
+                for (int i = 0; i < ps.length; i++) {
+                    if (!ps[i].getName().equals(expectedPNs[i])) {
+                        throw new RuntimeException("Fail: provider name " +
+                                "mismatch at index " + i + ", got " +
+                                ps[i].getName());
+                    }
+                }
+            } else {
+                throw new RuntimeException("Fail: # of providers mismatch");
+            }
+        }
+        System.out.println("=> Passed");
+    }
+
+
+    public static void main(String[] args)
+                throws NoSuchAlgorithmException {
+        String p = "SUN";
+        String key = "Signature.SHA256withDSA SupportedKeyClasses";
+        String valComp1 = "java.security.interfaces.DSAPublicKey";
+        String valComp2 = "java.security.interfaces.DSAPrivateKey";
+        String valComp2CN = valComp2.substring(valComp2.lastIndexOf('.') + 1);
+
+        // test using String filter
+        // 1. exact match
+        doit(key + ":" + valComp1 + "|" + valComp2, p);
+        // 2. value w/ space prefix
+        doit(key + ": " + valComp1, p);
+        // 3. value w/ space suffix
+        doit(key + ":" + valComp2 + " ", p);
+        // 4. partial value, e.g. class name only
+        doit(key + ":" + valComp2CN, p);
+        // 5. different values ordering
+        doit(key + ":" + valComp2 + "|" + valComp1, p);
+
+        // repeat above tests using filter Map
+        Map<String,String> filters = new HashMap<>();
+        filters.put(key, valComp1 + "|" + valComp2);
+        doit(filters, p);
+        filters.put(key, " " + valComp1);
+        doit(filters, p);
+        filters.put(key, valComp2 + " ");
+        doit(filters, p);
+        filters.put(key, valComp2CN);
+        doit(filters, p);
+        filters.put(key, valComp2 + " | " + valComp1);
+        doit(filters, p);
+
+        // add more filters to the map
+        filters.put("Signature.SHA256withDSA", "");
+        doit(filters, p);
+        filters.put("Cipher.Nonexisting", "");
```

Comment:
You mean the result should be empty even if it's not "Nonexisting"? Yes, the result is empty even if you change to use "AES" instead of "Nonexisting".